evm-ext/src/native_evm_context.rs

```diff
@@ -33,84 +32,49 @@ pub struct HeaderForExecution {
 #[derive(Tid)]
 pub struct NativeEVMContext<'a> {
     pub resolver: &'a dyn MoveResolver<PartialVMError>,
-    pub db: CacheDB<ResolverBackedDB<'a>>,
+    pub storage_trie: &'a dyn StorageTrieRepository,
     pub state_changes: Vec<HashMap<Address, Account>>,
     pub block_header: HeaderForExecution,
 }
 
 impl<'a> NativeEVMContext<'a> {
     pub fn new(
         state: &'a impl MoveResolver<PartialVMError>,
+        storage_trie: &'a impl StorageTrieRepository,
         block_header: HeaderForExecution,
     ) -> Self {
-        let inner_db = ResolverBackedDB::new(state);
-        let db = CacheDB::new(inner_db);
         Self {
             resolver: state,
-            db,
+            storage_trie,
             state_changes: Vec::new(),
             block_header,
         }
     }
 }
 
 pub struct ResolverBackedDB<'a> {
+    storage_trie: &'a dyn StorageTrieRepository,
     resolver: &'a dyn MoveResolver<PartialVMError>,
-    // This cache is used because each EVM account has a single resource for all
-    // its storage slots and therefore may be a large amount of data that takes
-    // a non-trivial amount of time to deserialize. By caching the storage representation
-    // in memory we only have to do the deserialization once per transaction. This
-    // optimization is likely helpful because it is common to access more than one
-    // storage slot for an EVM smart contract even in a single transaction. In the
-    // future if we choose to split storage across multiple resources to limit the
-    // size of a single resource, a cache will still be useful since reconstructing
-    // the storage from all the resource pieces will also be non-trivial.
-    //
-    // The cache must be wrapped in a type that allows interior mutability
-    // because the `DatabaseRef` interface uses immutable references. Since this
-    // cache is only used for a single transaction execution it is unlikely that
-    // it will ever need to multi-threaded access, so a thread-unsafe type like
-    // `RefCell` would be sufficient. But at the same time, I don't think the overhead
-    // of `RwLock` is that large, so I think it is ok to use.
-    storage_cache: RwLock<HashMap<Address, trie_types::AccountStorage>>,
```

Comment:
Just a side thought: any chance we would still benefit from having a cache of some sorts?

Reply:
I don't think so. This was closely tied to the Move resource serialization scheme. In that design, you had to deserialize the entire trie from a single byte array and ended up with a trie structure that's all in-memory. This approach tried to optimize the deserialization logic overhead, but in the end it was not enough.